presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java

```diff
@@ -1056,4 +1107,141 @@ public void append(JsonParser parser, BlockBuilder blockBuilder)
             blockBuilder.closeEntry();
         }
     }
+
+    private static class MapBlockBuilderAppender
+        implements BlockBuilderAppender
+    {
+        BlockBuilderAppender keyAppender;
+        BlockBuilderAppender valueAppender;
+        Type keyType;
+
+        MapBlockBuilderAppender(BlockBuilderAppender keyAppender, BlockBuilderAppender valueAppender, Type keyType)
+        {
+            this.keyAppender = keyAppender;
+            this.valueAppender = valueAppender;
+            this.keyType = keyType;
+        }
+
+        @Override
+        public void append(JsonParser parser, BlockBuilder blockBuilder)
+                throws IOException
+        {
+            if (parser.getCurrentToken() == JsonToken.VALUE_NULL) {
+                blockBuilder.appendNull();
+                return;
+            }
+
+            checkState(parser.currentToken() == START_OBJECT, "Expected a json object");
+            BlockBuilder entryBuilder = blockBuilder.beginBlockEntry();
+            int position = 0;
+            HashTable hashTable = new HashTable(keyType);
+            while (parser.nextValue() != END_OBJECT) {
+                keyAppender.append(parser, entryBuilder);
+                valueAppender.append(parser, entryBuilder);
+                if (hashTable.contains(entryBuilder, position)) {
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Duplicate keys are not allowed");
+                }
+                hashTable.add(entryBuilder, position);
+                position += 2;
+            }
+            blockBuilder.closeEntry();
+        }
+    }
+
+    // TODO: This class might be useful to other Map functions (transform_key, cast map to map, map_concat, etc)
+    // It is caller's responsibility to make the block data synchronized with the hash table
+    public static class HashTable
+    {
+        private static final int EXPECTED_ENTRIES = 20;
+        private static final float FILL_RATIO = 0.75f;
+        private static final int EMPTY_SLOT = -1;
+
+        private final Type type;
+
+        private int[] positionByHash;
+        private int hashCapacity;
+        private int maxFill;
+        private int hashMask;
+        private int size;
+
+        public HashTable(Type type)
+        {
+            this.type = requireNonNull(type, "type is null");
+            hashCapacity = arraySize(EXPECTED_ENTRIES, FILL_RATIO);
+            this.maxFill = calculateMaxFill(hashCapacity);
+            this.hashMask = hashCapacity - 1;
+            positionByHash = new int[hashCapacity];
+            Arrays.fill(positionByHash, EMPTY_SLOT);
+        }
+
+        public boolean contains(Block block, int position)
```

Comment:
If I understand correctly, the only way this HashTable can work is with a `BlockBuilder`. And since that `BlockBuilder` never changes, it should be passed in via the constructor.